odoo/addons/test_testing_utilities/tests/test_env.py

```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.tests.common import SavepointCase
+
+
+class TestEnv(SavepointCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestEnv, cls).setUpClass()
+        user = cls.env['res.users'].create({
+            'name': 'superuser',
+            'login': 'superuser',
+            'password': 'superuser',
+            'groups_id': [(6, 0, cls.env.user.groups_id.ids)],
+        })
+        cls.env = cls.env(user=user)
+        cls.cr = cls.env.cr
+
+        cls.env['res.company'].sudo().create({'name': 'Useless Company'})
+
+    def test_env_company_part_01(self):
+        """
+        There is not any assert in this test. The goal is actually to check the
+        values of the environment after the execution of this test
+        (see test_env_company_part_02)
+        """
+        company = self.env['res.company'].create({
+            "name": "Test Company",
+        })
+        self.env.user.write({
+            'company_id': company.id,
+            'company_ids': [(4, company.id), (4, self.env.company.id)],
+        })
+
+    def test_env_company_part_02(self):
+        user_env = self.env
+        sudo_env = self.env.user.sudo().env
```

Comment:
Is the environment `sudo_env` supposed to exist when `setUp()` is being run?
If that is the case, I suggest to assign it explicitly on the test class in `setUpClass()`, in order to guarantee that it is present in the weak set during `setUp()`.

Reply:
Yes, it even exists before the setUp, thanks to the `sudo` call in `setUpClass`. I will add a comment in the setup to highlight its creation

Reply:
It does not exist by necessity; I suspect it only exists because the garbage collector did not run between `setUpClass()` and `setUp()`.  That's the tricky bit about `WeakSet`: some elements might not always be there if they aren't explicitly referenced!